File: TP4-MVP/app/apps/flutter-app/lib/services/entrega_service.dart
import 'dart:convert';
import 'api_client.dart';
import '../models/entrega.dart';

class EntregaService {
  final ApiClient _api = ApiClient();

  Future<Entrega?> buscarPorTarefa(String tarefaId) async {
    final response = await _api.get('/entregas/tarefa/$tarefaId');
    if (response.statusCode == 404) return null;
    if (response.statusCode != 200) throw Exception('Erro ao buscar entrega');
    return Entrega.fromJson(jsonDecode(response.body));
  }

  Future<Entrega> submeter(String tarefaId, {String? comentarioAluno}) async {
    final response = await _api.post('/entregas/tarefa/$tarefaId', body: {
      if (comentarioAluno != null) 'comentarioAluno': comentarioAluno,
    });
    if (response.statusCode != 201) {
      final body = jsonDecode(response.body);
      throw Exception(body['error'] ?? 'Erro ao submeter entrega');
    }
    return Entrega.fromJson(jsonDecode(response.body));
  }

  Future<Entrega> avaliar(String entregaId, String feedbackOrientador, String statusAvaliacao) async {
    final response = await _api.patch('/entregas/$entregaId/avaliar', body: {
      'feedbackOrientador': feedbackOrientador,
      'statusAvaliacao': statusAvaliacao,
    });
    if (response.statusCode != 200) throw Exception('Erro ao avaliar entrega');
    return Entrega.fromJson(jsonDecode(response.body));
  }
}

File: TP4-MVP/app/apps/backend/src/application/services/entregaService.ts
import { prisma } from "../../infrastructure/prisma.js";

export const entregaService = {
  async submeter(tarefaId: string, usuarioId: string, data: { comentarioAluno?: string }) {
    const tarefa = await prisma.tarefa.findUnique({
      where: { id: tarefaId },
      include: { projeto: { select: { orientandoId: true, orientadorId: true } } },
    });
    if (!tarefa) throw Object.assign(new Error("Tarefa não encontrada"), { status: 404 });
    if (tarefa.projeto.orientandoId !== usuarioId) {
      throw Object.assign(new Error("Somente o orientando pode submeter entregas"), { status: 403 });
    }

    // Check if delivery already exists
    const existing = await prisma.entrega.findUnique({ where: { tarefaId } });
    if (existing) throw Object.assign(new Error("Já existe uma entrega para esta tarefa"), { status: 409 });

    return prisma.entrega.create({
      data: {
        tarefaId,
        comentarioAluno: data.comentarioAluno,
        statusAvaliacao: "PENDENTE",
      },
    });
  },

  async avaliar(entregaId: string, feedbackOrientador: string, statusAvaliacao: string) {
    const entrega = await prisma.entrega.findUnique({
      where: { id: entregaId },
      include: { tarefa: { include: { projeto: { include: { orientandoId: true, titulo: true } } } } },
    });
    if (!entrega) throw Object.assign(new Error("Entrega não encontrada"), { status: 404 });

    const updated = await prisma.entrega.update({
      where: { id: entregaId },
      data: { feedbackOrientador, statusAvaliacao: statusAvaliacao as any, dataAvaliacao: new Date() },
    });

    // Notificar orientando sobre feedback
    if (entrega.tarefa.projeto.orientandoId) {
      await prisma.notificacao.create({
        data: {
          usuarioId: entrega.tarefa.projeto.orientandoId,
          tipo: "FEEDBACK_ENTREGA",
          conteudo: `Feedback recebido para "${entrega.tarefa.titulo}": ${statusAvaliacao === "APROVADA" ? "Aprovada ✓" : "Necessita ajuste"}`,
        },
      });
    }

    return updated;
  },

  async buscarPorTarefa(tarefaId: string) {
    return prisma.entrega.findUnique({ where: { tarefaId } });
  },
};
